### DISCUSSION TITLE
Avalonia as replacement for libgdx?

### DISCUSSION THREAD

**Discussion by davidgiga1993** [AvaloniaUI/Avalonia#16992] [General]
A bit of background: We currently have a cross platform app based on libgdx (java) with approx. 400kLoC. Over the years the maintenance of some parts of libgdx causes us to seek alternative framework to ensure the longevity of the app (10+ years).
The app itself is not a game, although it exclusively uses custom UI controls for its specific domain, as well as partial handcrafted rendering to optimize battery lifetime of mobile devices. It's a unique use case which makes choosing the right framework tricky. On some screens there can be 600+ UI elements, a lot of them require a redraw each frame.

I played around a bit with avalonia and it ticks quite a few requirements. As migrating such a large codebase is not super trivial the goal would be to keep the custom rendering intact but replace raw opengl with the skia part of avalonia.
Effectively using one custom control in Avalonia and then calling the app specific scene graph.
(We're not using much of opengl's features, mostly rendering textures and 2D polygons, nothing 3D)

I'm looking for some input regarding rendering so make sure the base architecture somewhat fits:
- Is it possible to directly get a render call in the custom control without having to request it every frame (as done in the skia example in this repo)?
- How efficient is skia in rendering? The current code of the app does redraw the entire screen every frame, but adjusts the FPS dynamically to save power on low user interactions events. Is something similar possible?
- Anything else I should look out for when going this route that might interfere?

Thanks for your help.

**Comment by davidgiga1993** [comment]
Update: I did some testing rendering using the SKCanvas directly, however the performance seems pretty bad. I was under the assumption that Skia does batching automatically, so lots of calls do `DrawRect` for example would be batched together?

**Reply by kekekeks** [reply]
Are you drawing using sw rendering? It's rather slow. Follow the [custom composition visual example](https://github.com/AvaloniaUI/Avalonia/blob/282c0a60d8bf59cbf15af4dcd933ed9832c7cde6/samples/ControlCatalog/Pages/CompositionPage.axaml.cs#L191-L339) to get a stateful visual on our render thread and [skia api lease API example](https://github.com/AvaloniaUI/Avalonia/blob/a8a8060617d2669a6397f3b0acc43b5922dead1c/samples/RenderDemo/Pages/CustomSkiaPage.cs#L47). That should get you a gpu-accelerated SKCanvas directly attached to the render target with no extra overhead.

**Reply by davidgiga1993** [reply]
Thanks for your reply! I actually started with the skia lease api example, but drawing a couple of hundred rects (especially rounded rects) is very slow on mobile devices.

Update: And on my PC it actually showed about 10-14% gpu utilization on a rtx4090, so something seems wrong. 
The skia backend indicated opengl. Is there anything special that needs to be done for batching?

**Reply by davidgiga1993** [reply]
Did some more testing with the composition example app: On android (pixel 7 pro) it struggles to get constant 60fps when scrolling the left menu. The render times are ~6ms average which should be okay, so something else in the render pipelines causes this lag or the render time indicator is wrong(?). The phone is configured to 120Hz mode, so vsync shouldn't be the issue.
